### DISCUSSION TITLE
Easy loading of custom structs from RON files (bevy asset serde integration)

### DISCUSSION THREAD

**Discussion by inodentry** [bevyengine/bevy#1265] [Ideas]
I wish bevy offered an ergonomic way to load arbitrary structs and other data to be used in the game as files from the game assets.

Currently, AFAIK, one needs to write boilerplate to define a new custom asset type for each such piece of data / struct type that is to be loaded.

Some generic API for automagic integration with Serde, so that any `Deserialize` struct could be loaded as a bevy asset (ideally with any arbitrary `Deserializer`, not necessarily RON), would be awesome.

**Comment by cart** [comment]
Yeah I think this is a good idea. Imo this should be implemented as a "generic" AssetLoader that takes a Deserialize type, a set of file extensions, and some way to plug in an arbitrary serde Deserializer type. It would also need to take a Uuid, as all assets require a TypeUuid impl.

**Comment by inodentry** [comment]
Does bevy choose the `AssetLoader` to use for any given asset load simply based on the file extension?

**Reply by inodentry** [reply]
If the file extension is the only thing that can be used to differentiate different asset types, then does bevy take any care to prioritize longer / more specific extensions?

This stuff is important for the UX, since the goal is enable people to use the same file format (say, RON) for many structures of different types. For example, a game could have `.item.ron` for in-game items, `.level.ron` for level metadata, etc. Bevy should take care to prefer them when relevant, even if a `.ron` extension is also registered (say, for a generic RON asset loader).

**Reply by mockersf** [reply]
bevy's using https://doc.rust-lang.org/std/path/struct.Path.html#method.extension to get the extension, so only the last part

**Reply by cart** [reply]
Yeah the ambiguity is the problem and "multi-part" extensions would definitely help resolve that ambiguity. we could probably add support for that in bevy_asset, but its probably worth discussing the topic with @kabergstrom as we are in the process of integrating atelier_assets. The ambiguity could also be resolved in `my_asset.ron.meta` files (by manually specifying the asset type), but that would require manual action from users / wouldn't be very good ux.

**Comment by k0tran** [comment]
Take a look at [ron loader plugin](https://crates.io/crates/bevy_asset_ron), it seems pretty accurate

**Reply by van-sprundel** [reply]
Fun fact, she's the creator of that plugin!
Sadly it's archived
